src/middlewared/middlewared/plugins/afp.py

```diff
@@ -39,3 +42,214 @@ class Config:
+
+    async def validate_umask(self, data, schema_name):
+        verrors = ValidationErrors()
+
+        if not data.isdigit():
+                verrors.add(f"{schema_name}.umask",
+                            'The umask must be between 000 and 777.'
+                            )
+        else:
+            for i in range(len(data)):
+                umask_bit = int(data[i])
+                if umask_bit > 7 or umask_bit < 0:
+                    verrors.add(f"{schema_name}.umask",
+                                'The umask must be between 000 and 777.'
+                                )
+
+        if verrors:
+            raise verrors
+
+    @private
+    async def home_exists(self, home, schema_name, old=None):
+        verrors = ValidationErrors()
+        home_result = []
+
+        if home:
+            if old and old['id'] is not None:
+                id = old['id']
+
+                if not old['home']:
+                    # The user already had this set as the home share
+                    home_result = await self.middleware.call(
+                        'datastore.query', self._config.datastore,
+                        [('home', '=', True), ('id', '!=', id)],
+                        {'prefix': self._config.datastore_prefix})
+            else:
+                home_result = await self.middleware.call(
+                    'datastore.query', self._config.datastore,
+                    [('home', '=', True)],
+                    {'prefix': self._config.datastore_prefix})
+
+        if home_result:
+            verrors.add(f"{schema_name}.home",
+                        'Only one share is allowed to be a home share.')
+
+            raise verrors
+
+    @private
+    async def name_exists(self, data, schema_name, id=None):
+        verrors = ValidationErrors()
+        name = data['name']
+        path = data['path']
+        home = data['home']
+
+        if not name:
+            if home:
+                name = 'Homes'
+            else:
+                name = path.rsplit('/', 1)[-1]
+
+        if id is None:
+            name_result = await self.middleware.call(
+                'datastore.query', self._config.datastore,
+                [('name', '=', name)],
+                {'prefix': self._config.datastore_prefix})
+            path_result = await self.middleware.call(
+                'datastore.query', self._config.datastore,
+                [('path', '=', path)],
+                {'prefix': self._config.datastore_prefix})
+        else:
+            name_result = await self.middleware.call(
+                'datastore.query', self._config.datastore,
+                [('name', '=', name), ('id', '!=', id)],
+                {'prefix': self._config.datastore_prefix})
+            path_result = await self.middleware.call(
+                'datastore.query', self._config.datastore,
+                [('path', '=', path), ('id', '!=', id)],
+                {'prefix': self._config.datastore_prefix})
```

Comment:
Seems we are repeating ourselves a lot here just because of the filters, if I may suggest something:

```
name_filters =  [('name', '=', name)]
if id is not None:
    name_filters.append(('id', '!=', id))
```